File: app/src/main/java/com/xattacker/android/bingo/util/AlertDialogCreator.java
package com.xattacker.android.bingo.util;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.widget.ListAdapter;

import com.xattacker.android.bingo.R;

public final class AlertDialogCreator
{
	public enum AlertTitleType
	{
		NOTIFICATION_ALERT,
		WARNING_ALERT,
		ERROR_ALERT,
		CONFIRM_ALERT
	};

	public enum AlertButtonStyle
	{
		BUTTON_YES_NO,
		BUTTON_OK
	};

	// in order to hide constructor
	private AlertDialogCreator()
	{	
	}

	public static void showDialog
	(
	AlertTitleType aType, 
	String aText,
	Context aContext 
	)
	{
	   AlertDialog.Builder builder = createBuilder(aContext);
		builder.setTitle(createTitleText(aType, aContext));
		builder.setMessage(aText);
		  
		builder.setPositiveButton
		(
		aContext.getString(R.string.OK), 
		new DialogInterface.OnClickListener() 
		{  
			public void onClick(DialogInterface dialog, int which) 
			{  
				dialog.dismiss();
			}
		}
		);  
		
		builder.create();
		builder.show();
	}
	
	public static void showDialog
	(
	AlertTitleType aType, 
	AlertButtonStyle aStyle, 
	String aText,
	Context aContext,
	DialogInterface.OnClickListener aListener
	)
	{
		String button1 = null, button2 = null;
		
	   AlertDialog.Builder builder = createBuilder(aContext);
		builder.setTitle(createTitleText(aType, aContext));
		builder.setMessage(aText);
		
		switch (aStyle)
		{
			case BUTTON_YES_NO:
				button1 = aContext.getString(R.string.YES);
				button2 = aContext.getString(R.string.NO);
				break;
				
			case BUTTON_OK:
				button1 = aContext.getString(R.string.OK);
				break;
		}
		
		builder.setPositiveButton(button1, aListener); 
		
		if (button2 != null)
		{
			builder.setNegativeButton(button2, aListener); 
		}
		
		builder.create();
		builder.show();
	}
	
	public static void showDialog
	(
	String aTitle,
	ListAdapter aAdapter,
	Context aContext,
	DialogInterface.OnClickListener aListener
	)
	{
	   AlertDialog.Builder builder = createBuilder(aContext);
		builder.setTitle(aTitle);
		builder.setAdapter(aAdapter, aListener);
		
		builder.setPositiveButton
		(
		aContext.getString(R.string.CANCEL), 
		new DialogInterface.OnClickListener() 
		{  
			public void onClick(DialogInterface dialog, int which) 
			{  
				dialog.dismiss();
			}
		}
		);  
		
		builder.create();
		builder.show();
	}
	
	private static AlertDialog.Builder createBuilder(Context aContext)
	{
	   AlertDialog.Builder builder = new AlertDialog.Builder(aContext);
		builder.setIcon(R.drawable.app_icon);
		builder.setInverseBackgroundForced(true);
		
		return builder;
	}
	
	private static String createTitleText(AlertTitleType aType, Context aContext)
	{
		switch (aType)
		{
			case NOTIFICATION_ALERT:
				return aContext.getString(R.string.NOTIFICATION);

			case WARNING_ALERT:
				return aContext.getString(R.string.WARNING);

			case ERROR_ALERT:
				return aContext.getString(R.string.ERROR);
				
			case CONFIRM_ALERT:
				return aContext.getString(R.string.CONFIRM);
		}
		
		return "";
	}
}

File: app/src/main/java/com/xattacker/android/bingo/util/AlertDialogCreator.kt
package com.xattacker.android.bingo.util

import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.widget.ListAdapter
import com.xattacker.android.bingo.R

enum class AlertTitleType
{
    NOTIFICATION_ALERT(R.string.NOTIFICATION),
    WARNING_ALERT(R.string.WARNING),
    ERROR_ALERT(R.string.ERROR),
    CONFIRM_ALERT(R.string.CONFIRM);

    val value: Int

    constructor(value: Int)
    {
        this.value = value
    }
}

sealed class AlertButtonStyle
{
    open class TwiceButton(open val button1: Int, open val button2: Int): AlertButtonStyle()
    open class SingleButton(open val button: Int): AlertButtonStyle()

    // object
    object YesNo: TwiceButton(R.string.YES, R.string.NO)
    object OK: SingleButton(R.string.OK)
}


object AlertDialogCreator
{
    fun showDialog(aType: AlertTitleType, aText: String, aContext: Context)
    {
        val builder = createBuilder(aContext)
        builder.setTitle(aContext.getString(aType.value))
        builder.setMessage(aText)

        builder.setPositiveButton(aContext.getString(R.string.OK)) {dialog, which -> dialog.dismiss()}

        builder.create()
        builder.show()
    }

    fun showDialog(
        aType: AlertTitleType,
        aStyle: AlertButtonStyle,
        aText: String,
        aContext: Context,
        aClicked: (dialog: DialogInterface?, which: Int) -> Unit)
    {
        var button1: String? = null
        var button2: String? = null

        val builder = createBuilder(aContext)
        builder.setTitle(aContext.getString(aType.value))
        builder.setMessage(aText)

        when (aStyle)
        {
            is AlertButtonStyle.YesNo ->
            {
                button1 = aContext.getString(aStyle.button1)
                button2 = aContext.getString(aStyle.button2)
            }

            is AlertButtonStyle.OK -> button1 = aContext.getString(aStyle.button)
        }

        val listener = object : DialogInterface.OnClickListener
        {
            override fun onClick(dialog: DialogInterface?, which: Int)
            {
                aClicked.invoke(dialog, which)
            }
        }

        builder.setPositiveButton(button1, listener)

        if (button2 != null)
        {
            builder.setNegativeButton(button2, listener)
        }

        builder.create()
        builder.show()
    }

    fun showDialog(
        aTitle: String,
        aAdapter: ListAdapter,
        aContext: Context,
        aClicked: (dialog: DialogInterface?, which: Int) -> Unit)
    {
        val listener = object : DialogInterface.OnClickListener
        {
            override fun onClick(dialog: DialogInterface?, which: Int)
            {
                aClicked.invoke(dialog, which)
            }
        }

        val builder = createBuilder(aContext)
        builder.setTitle(aTitle)
        builder.setAdapter(aAdapter, listener)

        builder.setPositiveButton(aContext.getString(R.string.CANCEL)) {dialog, which -> dialog.dismiss()}

        builder.create()
        builder.show()
    }

    private fun createBuilder(aContext: Context): AlertDialog.Builder
    {
        val builder = AlertDialog.Builder(aContext)
        builder.setIcon(R.drawable.app_icon)
        builder.setInverseBackgroundForced(true)
        builder.setCancelable(false)

        return builder
    }
}
